### PR TITLE
fix: prevent idModel and qIdFunction from baseclass to be overwritten in digester which breaks multiple inheritance

### PR DESCRIPTION
**Pull request by manuelseeger**
Truthy check on array causes Idnames from grandparent type to be overwritten by parent type

### CODE REVIEW

**Comment by manuelseeger** [comment]
I added one more fix to the order the target classes are written to TS source, which gave me ts(2449) errors if the source EDMX doesn't define types in order of inheritance. 

It feels like a rather ugly fix. Looking for feedback and hopefully a better way to reorder. I am also not familiar enough with how the TS compiler works to say if the order of inheritance is a general problem or something only my setup has a problem with. 
I am also unclear why the TS compiler doesn't complain about interfaces referencing base types before those are declared, but does throw compile errors if classes reference base classes before those are declared.

**Comment by texttechne** [comment]
I've activated the test workflow for this PR.

File: packages/odata2ts/src/data-model/DataModelDigestion.ts [hunk 1]
```diff
@@ -221,7 +221,7 @@ export abstract class Digester<S extends Schema<ET, CT>, ET extends EntityType,
         }
 
         props.push(...baseModel.props);
-        if (baseModel.keyNames) {
+        if (baseModel.keyNames.length) {
```
Comment (texttechne) [inline packages/odata2ts/src/data-model/DataModelDigestion.ts:224]: should be `baseModel.keyNames?.length` => this is the test failure

**Comment by texttechne** [comment]
Actually, I don't quite get what you're trying to solve here.
Do you have an example? What's the source EDMX and what't the exepcected TS interface you expect? Especially regarding the ordering you mentiion an example would be really helpful.

And then we need some tests. You could try to add one in the [DataModelDigestion Tests](https://github.com/odata2ts/odata2ts/blob/main/packages/odata2ts/test/data-model/DataModelDigestionTests.ts). You cannot execute those tests directly as they are shared for the v2 and the v4 version of the digester. So you actually execute one of those:
* [v2/DataModelDigestion.test.ts](https://github.com/odata2ts/odata2ts/blob/main/packages/odata2ts/test/data-model/v2/DataModelDigestion.test.ts)
* [v4/DataModelDigestion.test.ts](https://github.com/odata2ts/odata2ts/blob/main/packages/odata2ts/test/data-model/v4/DataModelDigestion.test.ts)

**Comment by manuelseeger** [comment]
Sure, I am trying to build this service: 
[cpi.xml.txt](https://github.com/odata2ts/odata2ts/files/11240482/cpi.xml.txt)

with 

```
    cpi: {
      source: "int-test/fixture/v2/cpi.xml",
      output: "build/v2/cpi",
      allowRenaming: true,
      propertiesByName: [
        {
          name: "FromId_",
          mappedName: "fromIdPrivate",
        },
      ],
    },
```

I ran into 2 issues: 
1. grandparent IDs get overwritten by parent IDs (services will try to import Id types that don't exist)
2. Qobjects reference base classes before these classes are declared

I'll spend more time on it though this might have been too quick an attempt.

**Comment by texttechne** [comment]
Ok, I've understood both bugs now. And because I've never said it: Big thanks for the fixes, much appreciated!

Can you make two PRs out of this? So one PR for each fix.
Sorry for the inconvenience, but it's much clearer and cleaner then. In the end I will squash each PR so that each PR will result in 1 commit. Hope this makes sense.

The first bugfix is fine already. You perfectly extended the tests for this: as minimal as possible!
I would suggest to keep it in here and refactor the other bugfix into the next PR.

Regarding the second bugfix. I think you can and should reuse the existing test in `v4/EntityBasedGenerationTests.ts` which you copied over to v2. Just make the changes in there including your comment.
There's nothing special when it comes to V4 or V2 in regards to entity inheritance: works exactly the same for both versions. But since I had to put these tests somewhere they landed in the V4 tests.

Additionally, when using the `EntityBasedGenerationTests.ts` you also test the model generation. And while its not a bug in TS to extend an interface which is only defined later, your proposed ordering should also apply here. The ordering for the model generation should be stable and in line with the q-object generation.

**Comment by manuelseeger** [comment]
Got it. This is for ids of grandparents being overwritten then.

**Pull request merged**